File: 2024/17/main.cpp

#include "../helpers/extras.hh"

#include <cmath>

void getData(const char* filename, size_t* A, size_t* B, size_t* C, std::vector<char>& program)
{
    std::ifstream file(filename);
    std::string line;

    // Register A
    getline(file, line);
    *A = std::atol(splitString(line, ":")[1].c_str());

    // Register B
    getline(file, line);
    *B = std::atol(splitString(line, ":")[1].c_str());

    // Register C
    getline(file, line);
    *C = std::atol(splitString(line, ":")[1].c_str());

    // Skip the next line
    getline(file, line);

    // Program
    getline(file, line);
    std::vector<std::string> programStr = splitString(splitString(line, ":")[1],",");

    for (auto op:programStr)
        program.push_back(std::atoi(op.c_str()));
    file.close();
}

size_t combo_operand(char value, size_t A, size_t B, size_t C)
{
    switch (value) {
        case 0:
        case 1:
        case 2:
        case 3:
            return value;
        case 4:
            return A;
        case 5:
            return B;
        case 6:
            return C;
    }
    return 0;
}

std::vector<char> run(size_t A, size_t B, size_t C, std::vector<char>& program)
{
    size_t pointer = 0;

    std::vector<char> outputs;

    while (pointer < program.size()) {
        char opcode = program[pointer];
        char operand = program[pointer+1];

        switch (opcode) {
            case 0:
                A/=std::pow(2, combo_operand(operand, A, B, C));
                break;
            case 1:
                B ^= operand;
                break;
            case 2:
                B = combo_operand(operand, A, B, C) % 8;
                break;
            case 3:
                if (A != 0) {
                    pointer = operand;
                    continue;
                }
                break;
            case 4:
                B ^= C;
                break;
            case 5:
                outputs.push_back(combo_operand(operand, A, B, C) % 8);
                break;
            case 6:
                B = A / std::pow(2, combo_operand(operand, A, B, C));
                break;
            case 7:
                C = A / std::pow(2, combo_operand(operand, A, B, C));
                break;
        }
        pointer += 2;
    }
    return outputs;
}

void doPart1(const char* filename)
{
    size_t A,B,C;
    std::vector<char> program;
    getData(filename, &A, &B, &C, program);

    std::vector<char> output = run(A, B, C, program);
    printf("%s\n", joinString(output, ',').c_str());
}

void doPart2(const char* filename)
{
    size_t A,B,C;
    std::vector<char> program;
    if (std::string(filename) == "example.dat") {
        A = 2024;
        B = 0;
        C = 0;

        program.push_back(0);
        program.push_back(3);
        program.push_back(5);
        program.push_back(4);
        program.push_back(3);
        program.push_back(0);

    } else {
        getData(filename, &A, &B, &C, program);
    }

    // Get initial A
    A = std::pow(8,(program.size()-1));

    size_t numUpdates = 0;
    while (true) {
        std::vector<char> output = run(A, B, C, program);

        // May be the same, now check values
        bool isSame = true;
        for (size_t k = 0; k < program.size(); k++) {
            isSame &= (output[k] == program[k]);
        }

        if (isSame) {
            #ifdef DEBUG
                printf("%4ld: %16ld -> %s\n", numUpdates, A, joinString(output,',').c_str());
            #else
                printf("%ld\n",A);
            #endif
            break;
        }

        size_t add = 0;
        for (size_t k = 0; k < program.size(); k++) {
            const char o = output[output.size()-1-k];
            const char p = program[program.size()-1-k];
            if (o != p) {
                add = std::pow(8, program.size()-1-k);
                #ifdef DEBUG
                    printf(
                        "%4ld: %16ld -> %s :: adding %ld(8**%ld)\n",
                        numUpdates, A, joinString(output,',').c_str(), add, program.size()-1-k
                    );
                #endif
                numUpdates++;
                A += add;
                break;
            }
        }
    }
}

int main(int narg, char* args[])
{
    std::string filename = "example.dat";
    if (narg > 1) filename = std::string(args[1]);
    doPart1(filename.c_str());
    doPart2(filename.c_str());
    return 0;
}

File: 2024/17/main.py
#!/usr/bin/env python3

import sys

def run_program(registers, program):
    def combo_operand(value):
        match value:
            case 0 | 1 | 2 | 3:
                return value
            case 4:
                return A
            case 5:
                return B
            case 6:
                return C

    A, B, C = registers
    pointer = 0
    outputs = []

    while pointer < len(program):
        opcode = program[pointer]
        operand = program[pointer + 1]

        if opcode == 0:
            A //= 2 ** combo_operand(operand)
        elif opcode == 1:
            B ^= operand
        elif opcode == 2:
            B = combo_operand(operand) % 8
        elif opcode == 3:
            if A != 0:
                pointer = operand
                continue
        elif opcode == 4:
            B ^= C
        elif opcode == 5:
            outputs.append(combo_operand(operand) % 8)
        elif opcode == 6:
            B = A // (2 ** combo_operand(operand))
        elif opcode == 7:
            C = A // (2 ** combo_operand(operand))

        pointer += 2

    return outputs

def doPart1(filename:str):
    registers, program = open(filename,"r").read().strip().split("\n\n")
    registers = [int(x.split(":")[1]) for x in registers.strip().split("\n")]
    program = [int(inst) for inst in program.strip().split(":")[1].strip().split(",")]
    result = run_program(registers, program)
    print(",".join(map(str, result)))

def doPart2(filename:str):
    if filename == "example.dat":
        registers = [2024,0,0]
        program = [0,3,5,4,3,0]
    else:
        registers, program = open(filename,"r").read().strip().split("\n\n")
        registers = [int(x.split(":")[1]) for x in registers.strip().split("\n")]
        program = [int(inst) for inst in program.strip().split(":")[1].strip().split(",")]

    A = 8**(len(program)-1)

    while True:
        result = run_program([A, 0, 0], program)

        if result == program:
            print(A)
            break

        add = 0
        for i in range(len(result) - 1, -1, -1):
            if result[i] != program[i]:
                add = 8**i
                A += add
                break

def main():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    doPart1(filename)
    doPart2(filename)

if __name__ == "__main__":
    main()
